<a href="https://colab.research.google.com/github/KhytsenkoIgor/python-data-analysis-tasks/blob/main/hw_10_3_%D0%A0%D0%BE%D0%B7%D1%88%D0%B8%D1%80%D0%B5%D0%BD%D1%96_%D0%BC%D0%B5%D1%82%D0%BE%D0%B4%D0%B8_%D0%BE%D0%B1%D1%80%D0%BE%D0%B1%D0%BA%D0%B8_%D0%B4%D0%B0%D0%BD%D0%B8%D1%85_%D1%83_Pandas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнє завдання до теми "Розширені методи обробки даних у Pandas"

В цьому домашньому завданні продовжуємо працювати з набором даних `supermarket_sales.csv`.

0. Імпортуйте бібліотеку pandas та зчитайте дані у змінну `df` типу `pandas.DataFrame`.

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
df = pd.read_csv('drive/MyDrive/Модуль_10/data/supermarket_sales.csv')

1. Відсортуйте дані за спаданням ціни товара ('Unit price') і виведіть 5 перших рядочків на екран.

In [ ]:
df.sort_values('Unit price', ascending=False).head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
983,148-41-7930,C,Naypyitaw,Normal,Male,Health and beauty,99.96,7,34.986,734.706,1/23/2019,10:33,Cash,699.72,4.761905,34.986,6.1
122,219-22-9386,B,Mandalay,Member,Male,Sports and travel,99.96,9,44.982,944.622,3/9/2019,17:26,Credit card,899.64,4.761905,44.982,4.2
930,641-62-7288,B,Mandalay,Normal,Male,Home and lifestyle,99.92,6,29.976,629.496,3/24/2019,13:33,Ewallet,599.52,4.761905,29.976,7.1
494,437-53-3084,B,Mandalay,Normal,Male,Fashion accessories,99.89,2,9.989,209.769,2/26/2019,11:48,Ewallet,199.78,4.761905,9.989,7.1
283,667-92-0055,A,Yangon,Member,Male,Health and beauty,99.83,6,29.949,628.929,3/4/2019,15:02,Ewallet,598.98,4.761905,29.949,8.5


2. Знайдіть інвойс (рядок в даних) в категорії товарів `Electronic accessories`, в якому був куплений товар з найбільшою кількістю і найбільшою ціною. Виведіть цей рядок даних на екран.

In [ ]:
electronic = df[df['Product line'] == 'Electronic accessories']


In [ ]:
electronic[electronic['Quantity'] == electronic['Quantity'].max()].nlargest(1, 'Unit price')

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3


**Перевірка :**

In [ ]:
electronic.sort_values(['Quantity', 'Unit price'],ascending=[False, False]).head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating
105,704-48-3927,A,Yangon,Member,Male,Electronic accessories,88.67,10,44.335,931.035,1/12/2019,14:50,Ewallet,886.7,4.761905,44.335,7.3
988,267-62-7380,C,Naypyitaw,Member,Male,Electronic accessories,82.34,10,41.170,864.570,3/29/2019,19:12,Ewallet,823.4,4.761905,41.170,4.3
109,861-77-0145,C,Naypyitaw,Member,Male,Electronic accessories,81.97,10,40.985,860.685,3/3/2019,14:30,Cash,819.7,4.761905,40.985,9.2
457,533-33-5337,B,Mandalay,Normal,Male,Electronic accessories,79.39,10,39.695,833.595,2/7/2019,20:24,Cash,793.9,4.761905,39.695,6.2
314,227-07-4446,C,Naypyitaw,Member,Female,Electronic accessories,78.13,10,39.065,820.365,2/10/2019,20:51,Cash,781.3,4.761905,39.065,4.4


**Відповідь:**
Invoice з найбільшою кількістю товару та найбільшою ціною в категорії Electronic accessories має ID 704-48-3927.

3.1. Створіть нову колонку в даних `Rating Category`, яка визначає категорію рейтингу (`Rating`): високий, середній чи низький за наступним правилом:
- якщо рейтинг вищий або рівний 8 - то він високий (`High`)
- якщо рейтинг від 6 (включно) до 8 (не включно) - то він середній (`Medium`)
- інакше - низький (`Low`).

In [ ]:
def rating_category(rating):
    if rating >= 8:
        return 'High'
    elif rating >= 6:
        return 'Medium'
    else:
        return 'Low'


In [ ]:
df['Rating Category'] = df['Rating'].apply(rating_category)

**Перевірка:**

In [ ]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low


In [ ]:
df[['Rating', 'Rating Category']].head(10)

,Rating,Rating Category
0,9.1,High
1,9.6,High
2,7.4,Medium
3,8.4,High
4,5.3,Low
5,4.1,Low
6,5.8,Low
7,8.0,High
8,7.2,Medium
9,5.9,Low


3.2. Виведіть, який відсоток даних становить кожне зі значень категорій рейтингу.

In [ ]:
(df['Rating Category'].value_counts(normalize=True)* 100 )

,proportion
Rating Category,
Medium,35.6
High,32.9
Low,31.5


4.1. Запустіть код нижче аби створити змінну `time`. Ця змінна містить час в форматі "година:хвилина". Напишіть функцію `get_hour`, яка буде зі змінної формату як `time` дістати лише години і конвертувати їх у тип `int`.

Запустіть написану функцію на змінній `time` та перевірте правильність роботи функції.

In [ ]:
time = '11:28'

In [ ]:
def get_hour(time):
    return int(time[:2])

In [ ]:
get_hour(time)

11

4.2. З допомогою функції `get_hour` і метода `DataFrame.apply()` створіть нову колонку `Hour` в нашому наборі даних, яка буде містити години здійснення покупки, обчислені на основі часу здійснення покупки `Time`.

In [ ]:
df['Hour'] = df['Time'].apply(get_hour)

**Перевірка:**

In [ ]:
df.head()

,Invoice ID,Branch,City,Customer type,Gender,Product line,Unit price,Quantity,Tax 5%,Total,Date,Time,Payment,cogs,gross margin percentage,gross income,Rating,Rating Category,Hour
0,750-67-8428,A,Yangon,Member,Female,Health and beauty,74.69,7,26.1415,548.9715,1/5/2019,13:08,Ewallet,522.83,4.761905,26.1415,9.1,High,13
1,226-31-3081,C,Naypyitaw,Normal,Female,Electronic accessories,15.28,5,3.8200,80.2200,3/8/2019,10:29,Cash,76.40,4.761905,3.8200,9.6,High,10
2,631-41-3108,A,Yangon,Normal,Male,Home and lifestyle,46.33,7,16.2155,340.5255,3/3/2019,13:23,Credit card,324.31,4.761905,16.2155,7.4,Medium,13
3,123-19-1176,A,Yangon,Member,Male,Health and beauty,58.22,8,23.2880,489.0480,1/27/2019,20:33,Ewallet,465.76,4.761905,23.2880,8.4,High,20
4,373-73-7910,A,Yangon,Normal,Male,Sports and travel,86.31,7,30.2085,634.3785,2/8/2019,10:37,Ewallet,604.17,4.761905,30.2085,5.3,Low,10


In [ ]:
df[['Time', 'Hour']].head(7)

,Time,Hour
0,13:08,13
1,10:29,10
2,13:23,13
3,20:33,20
4,10:37,10
5,18:30,18
6,14:36,14


4.3. Виведіть кількість значень за кожною годиною покупки і дайте відповідь "О яких годинах найбільше здійснюють покупок покупці і чому на Ваш погляд саме о цих годинах? Чи це слогічно, що найбільше покупок саме в ці години?".

In [ ]:
df['Hour'].value_counts()

,count
Hour,
19,113
13,103
15,102
10,101
18,93
11,90
12,89
14,83
16,77


**Відповідь:**  Найбільше покупок покупці здійснюють в такі години:

* 19 год. (113 покупок) на мій погляд це логічно,тому що вечером люди часто роблять покупки після роботи прямуючи до дому;
* 13 та 15 година ( 103 та 102 покупки)за рейтингом наступна найбільша кількість покупок,скоріше за все це пов'язано з перервою на обід.


5. Виведіть середнє значення чеку (`Total`) для покупок в містах, наявних в даних, за допомогою функції `groupby`.

In [ ]:
df.groupby('City')['Total'].mean().round(2)

,Total
City,
Mandalay,319.87
Naypyitaw,337.10
Yangon,312.35


6. Виведіть кількість покупок, середнє, мінімальне та максимальне значення для загальної суми чеку (Total) в містах, наявних в даних, за допомогою функції groupby.
Зробіть висновки з цих показників та дайте рекомендації для мережі супермаркетів.

In [ ]:
df.groupby('City')['Total'].agg(['count', 'mean', 'min', 'max']).round(2)

,count,mean,min,max
City,,,,
Mandalay,332,319.87,18.64,1022.49
Naypyitaw,328,337.10,10.68,1042.65
Yangon,340,312.35,12.69,1039.29


**Відповідь:** *Найбільша кількість покупок у місті Yangon 340,0 але середній чек тут найнижчий 312,35. У Naypyitaw покупок трохи менше  328, проте саме тут найбільший середній чек 337,10 та максимальний чек  1042,65.Місто Mandalay займає середню позицію за цими показниками. На мою думку, у місті Yangon можна спробувати збільшити середній чек за допомогою акцій або пропозицій на додаткові товари. Для міста Naypyitaw варто зберігати поточний підхід, оскільки покупці в середньому витрачають більше.*  



7. Ми внесли деякі зміни в наш набір даних, тож, давайте запишемо результат у файл. Запишіть оновлену версію даних у файл `supermarket_sales_enriched.csv`. Зверніть увагу, аби не записати нічого зайвого, наприклад, індекси рядків нам в файлі не потрібні, оскільки вони стандартні.

In [ ]:
df.to_csv('supermarket_sales_enriched.csv', index=False)